cpp/include/raft/neighbors/ivf_pq.cuh

```diff
@@ -75,6 +75,48 @@ auto build(raft::device_resources const& handle,
   return detail::build(handle, params, dataset, n_rows, dim);
 }
 
+/**
+ * @brief Decode `n_take` consecutive records of a single list (cluster) in the compressed index
+ * starting at given offset `n_skip`.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   // We will reconstruct the fourth cluster
+ *   uint32_t label = 3;
+ *   // Get the list size
+ *   uint32_t list_size = 0;
+ *   raft::copy(&list_size, index.list_sizes().data_handle() + label, 1, res.get_stream());
+ *   res.sync_stream();
+ *   // allocate the buffer for the output
+ *   auto decoded_vectors = raft::make_device_matrix<float>(res, list_size, index.dim());
+ *   // decode the whole list
+ *   ivf_pq::reconstruct_list_data(res, index, decoded_vectors.view(), label, 0);
+ * @endcode
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices in the source dataset
+ *
+ * @param[in] res
+ * @param[in] index
+ * @param[out] out_vectors
+ *   the destination buffer [n_take, index.dim()].
+ *   The length `n_take` defines how many records to reconstruct,
+ *   it must be smaller than the list size.
+ * @param[in] label
+ *   The id of the list (cluster) to decode.
+ * @param[in] n_skip
+ *   How many records in the list to skip.
+ */
+template <typename T, typename IdxT>
+void reconstruct_list_data(raft::device_resources const& res,
+                           const index<IdxT>& index,
+                           device_matrix_view<T, uint32_t, row_major> out_vectors,
+                           uint32_t label,
+                           uint32_t n_skip)
```

Comment:
@cjnolet , could you please check if this interface is suitable for integration with FAISS?

Reply:
Also don't you need something like this interface to be able to convert our GPU index to their CPU index type (more-or-less efficiently, converting one cluster at a time)?

Reply:
So, if I understand this correctly, the current proposed API here covers the CodePacker needs, right?

Reply:
If this is intended to cover the code packer's needs, I'd like it if we adjusted the verbiage there, and maybe even use the verbiage from the code packer itself (e.g. `pack_single`, `unpack_all`) because the two use-cases are pretty different. Naming aside, I think this *could* work for the `unpack` side. We will still need the version that goes the other way, though.

Reply:
As for the second task, I'm not sure how we're supposed to accomplish it. As I said, we don't sort input indices in the database, so locating a single id in the database would entail a full linear scan. We can get some speedup thanks to batching, but that still sounds really slow. Do you know if it is the same in faiss?

Reply:
> We can get some speedup thanks to batching, but that still sounds really slow. Do you know if it is the same in faiss?

Maybe I've misunderstood how we handle being given indexes today. What happens today when someone calls `ivf_pq::extend` for the following 3 indexes: `[500, 2, 50]`? Let's say these 3 are the only 3 indexes in the database.

Reply:
At the moment, we store the cluster data and indices in `ivf::list` structures for each cluster separately. That is, you can get a list of source indices for each cluster. In your example, with two clusters we may end up with two index lists `[[500], [2, 5]]`. To complicate things further,  `ivf_pq::extend`  writes vectors/indices in an undefined order for performance reasons (concurrent `atomicAdd` involved).

As a result we cannot rely in any way on the order in which the data is written. That being said, we can sort all indices by their source id in a "preparation" phase for a given index, so that the indices can be reconstructed faster afterwards.

Btw, the same issue applies to ivf-flat. Luckily, the interleaved layout does not affect the indices in both implementations, they are stored flat (per-cluster).

Reply:
> At the moment, we store the cluster data and indices in ivf::list structures for each cluster separately.

Okay, so you are saying that the indices provided from the user are directly stored in the sparse structure representing the IVF lists? So, for example, the index 500 would still be associated w/ the first (and only) vector in the  first cluster (going from your example cluster assignments)? If this is the case, you are right that we would need to use a hash table or sort and use a binary search to look this up. At least that tells me it could be possible, nonetheless. 

If, on the other hand, we are normalizing the given index so that it's linear with all the other indexes then I think all bets are off.

Reply:
Yes, basically, we have to store the user indices, because this is what we return from the `ivf_*::search` along with the distances. Hence we can create a mapping {user index -> (cluster, incluster_offset)} and use it for faster retrieval by user index. The only problem is that the time complexity of creating it would be something like O(n log n) and it would require at least `(sizeof(IdxT) + 8) * n` extra storage for the struct itself, both of which can be a lot if we're talking about billion-record ivf-pq databases.

Reply:
I'd suggest I enhance the interface to accept (cluster, list of in-cluster indices/offsets) in addition to the current one in this PR. Then in a follow-up we can add this index mapping struct directly to `ivf::list` module so it can be shared between ivf-flat and ivf-pq.

Reply:
Sorry, @achirkin I'm still kind of stuck on the CodePacker problem because that's the blocker at the moment the integration. Can you explain how this interface allows us to implement the CodePacker interface for RAFT's IVFPQ methods? Im having trouble figuring out where the information (like block size, etc...) is specified. Is that a constant? 

In the end, we are going to have to fill that in for RAFT in order to convert to/from FAISS cpu/gpu indexes. The FAISS folks have asked us to use the CodePacker API instead of the reconstruct_n API. It sounds like we are almost there but I think there might be a gap in my understanding of your proposed interface.

Reply:
Yes, they are using the per-block flat vectors as an intermediate format for conversion. So, for example, we should be able to loop through each vector or block and write its codes in the flat form using one code packer implementation and then immediately unpack from flat to another code packer implementation- they provide APIs for doing 1, many, and all. If we do this right, it should be super easy to convert to/from FAISS existing gpu formats in addition to the cpu formats.

Reply:
Update.

I've implemented `unpack_list_data` so far with a small change in the logic. I always write one PQ code per byte, ignoring `pq_bits` (faiss does not support smaller code sizes). Hence the constants change a little bit:
```
nvec == list.values.extent(2) == kIndexGroupSize;
code_size == pq_dim;
block_size ==  list.values.extent(1) *  list.values.extent(2) * list.values.extent(3)
```

NB: it's not a problem to pack the flat codes using `pq_bits` bits per code, but I figure there's simply no use for it; if you think otherwise, we can make this change.

Reply:
@cjnolet I've updated the public API (implementation follows soon), could you have a look?

Note, although the public api provides now a way to read raw flat PQ codes, it's not suitable for the CodePacker, because it's not raw pointers. However, one of the corresponding functions in `detail` namespace hopefully should work for this purpose.

Reply:
The purpose of these pack and unpack APIs is so that we can call them through FAISS. We can't invoke anything below the public APIs external to RAFT, so we need to expose these publicly. Exposing them in a `helper` namespace further highlights to the user that these aren't required for normal operation but provide additional functionality that might be desired. 

In order to integrate w/ FAISS, we need to be able to provide a (stable) contract to them so that we can help them convert between different index types through these codepacker APIs. I agree that how these pack/unpack functions ultimately work "under the hood" should be hidden behind `detail` but the functions themselves to perform the pack/unpack, should be publicly accessable. That says to the world "We will not change these APIs, nor the way they function, without deprecation first". 

I'm only suggesting that we use `raft::span` in the case where we don't feel `mdspan` accurately represents a type. If an mdspan vector can represent this, that's great!

Reply:
> That says to the world "We will not change these APIs, nor the way they function, without deprecation first".

Oh, now I see what you mean. I agree, we should export `detail::pack_list_data` and `detail::unpack_list_data` into a public namespace, to make sure we don't change their signature and break faiss integration by accident. I'm also not opposed to the idea of moving the construct/reconstruct and other similar functions to the `ivf_pq::helpers` public namespace.
The only thing that bothers me, is that the overloads we need for the codepacker go really against the current design of the public api. Maybe, as an exception, we export these two into something like `ivf_pq::codepacker` to discourage other users from using them? Or even implement the whole codepacker api in raft in that namespace?

Reply:
I have added a public API for pack_list_data.

Reply:
Note, for the codepacker, we'd probably need to expose this overload https://github.com/rapidsai/raft/blob/926f510aab3ec16844b9b18b4c04fcc51e8f1f7d/cpp/include/raft/neighbors/detail/ivf_pq_build.cuh#L575-L580

Reply:
I have missed this message before. I have  exposed the pack/unpack versions  that take list_data directly under `ivf_pq::helpers::codepacker` namespace.

In our implementation we still need to provide the `pq_bits` parameter. Otherwise it is very similar to the CodePacker API.

One question: what does `block` refer to in CodePacker? In our API it refers to a set of interleaved blocks, which can be the data for a whole cluster, or less.